torch/onnx/utils.py

```diff
@@ -1245,42 +1240,68 @@ def export_to_pretty_string(
 
     Returns:
-        Tuple[torch._C.Graph, List[str]], where the list includes the names
-        of the unconvertible ops.
+        The JIT graph and a list of unconvertible ops in the format of "domain::op".
     """
 
     opset_version = opset_version or _constants.ONNX_DEFAULT_OPSET
     symbolic_helper._set_opset_version(opset_version)
-    # operator_export_type is set to ONNX_FALLTHROUGH by default so that if an op is not supported
-    # in ONNX, fall through will occur and export the operator as is, as a custom ONNX op.
-    with exporter_context(model, training, False):
-        args = _decide_input_format(model, args)
-        graph, params_dict, torch_out = _model_to_graph(
-            model,
-            args,
-            # So that if an op connot be converted to ONNX, it will be kept
-            # as-is rather than cause a failure.
-            operator_export_type=_C_onnx.OperatorExportTypes.ONNX_FALLTHROUGH,
-        )
-    unsupported_ops = list()
-    supported_namespaces = {"onnx", "prim", "quantized"}
+
+    try:
+        with exporter_context(model, training, verbose=False):
+            # Create a mostly clean JIT graph that contains the plain aten and
+            # other ops we can check with the symbolic registry.
+            # NOTE: We don't want to actually convert any ops to ONNX or run any
+            # symbolic functions because there is a higher chance that a pass
+            # fails or an unconvertible op messes up the graph during ONNX conversion.
+            # This way we can always generate a list just by looking at the names
+            # of the ops in the graph.
+            args = _decide_input_format(model, args)
+            model = _pre_trace_quant_model(model, args)
+            graph, _, _, module = _create_jit_graph(model, args)
+            _C._jit_pass_inline(graph)
+            _C._jit_pass_onnx_remove_inplace_ops_for_onnx(graph, module)
+            _C._jit_pass_erase_number_types(graph)
+            _C._jit_pass_dce_allow_deleting_nodes_with_side_effects(graph)
+    except Exception as e:
+        raise errors.OnnxExporterError(
+            "Failed to discover unconvertible ops because of errors during the JIT graph "
+            "generation process."
+        ) from e
+
+    unsupported_ops = []
     for node in graph.nodes():
-        if node.kind().split(":")[0] not in supported_namespaces:
-            unsupported_ops.append(node.kind())
+        domain_op = node.kind()
+        if domain_op.startswith("onnx::") or domain_op.startswith("prim::"):
+            # We consider onnx and prim ops as supported ops, even though some "prim"
+            # ops are not implemented as symbolic functions, because they may be
+            # eliminated in the conversion passes. Users may still see errors caused
+            # by prim ops even though they don't show up in the list.
+            continue
+        if not registration.registry.is_registered_op(domain_op, opset_version):
```

Comment:
I am curious if we can simply use torch graph only with registry to check unconvertible ops?

Reply:
Yeah. This is the point where we have access to the ops where the graph is flattened.